src/node_report.cc

```diff
@@ -430,126 +435,201 @@ void TriggerNodeReport(Isolate* isolate, DumpEvent event, const char* message, c
 #else
       snprintf(pathname, sizeof(pathname), "%s%s%s", report_directory, "/", filename);
 #endif
-      fp = fopen(pathname, "w");
+      outfile.open(pathname, std::ios::out);
     } else {
-      fp = fopen(filename, "w");
+      outfile.open(filename, std::ios::out);
     }
     // Check for errors on the file open
-    if (fp == nullptr) {
+    if (!outfile.is_open()) {
       if (strlen(report_directory) > 0) {
-        fprintf(stderr, "\nFailed to open Node.js report file: %s directory: %s (errno: %d)\n", filename, report_directory, errno);
+        std::cerr << "\nFailed to open Node.js report file: " << filename << " directory: " << report_directory << " (errno: " << errno << ")\n";
       } else {
-        fprintf(stderr, "\nFailed to open Node.js report file: %s (errno: %d)\n", filename, errno);
+        std::cerr << "\nFailed to open Node.js report file: " << filename << " (errno: " << errno << ")\n";
       }
       return;
     } else {
-      fprintf(stderr, "\nWriting Node.js report to file: %s\n", filename);
+      std::cerr << "\nWriting Node.js report to file: " << filename << "\n";
     }
   }
 
+  // Pass our stream about by reference, not by copying it.
+  std::ostream &out = outfile.is_open() ? outfile : *outstream;
+
+  WriteNodeReport(isolate, event, message, location, filename, out, &tm_struct);
+
+  // Do not close stdout/stderr, only close files we opened.
+  if(outfile.is_open()) {
+    outfile.close();
+  }
+
+  std::cerr << "Node.js report completed\n";
+  if (name != nullptr) {
+    snprintf(name, NR_MAXNAME + 1, "%s", filename);  // return the NodeReport file name
+  }
+
+}
+
+void GetNodeReport(Isolate* isolate, DumpEvent event, const char* message, const char* location, std::ostream& out) {
+  // Obtain the current time and the pid (platform dependent)
+  TIME_TYPE tm_struct;
+#ifdef _WIN32
+  GetLocalTime(&tm_struct);
+#else  // UNIX, OSX
+  struct timeval time_val;
+  gettimeofday(&time_val, nullptr);
+  localtime_r(&time_val.tv_sec, &tm_struct);
+#endif
+  WriteNodeReport(isolate, event, message, location, nullptr, out, &tm_struct);
+}
+
+static void walkHandle(uv_handle_t* h, void* arg) {
+  std::string type;
+  std::string data = "";
+  std::ostream* out = reinterpret_cast<std::ostream*>(arg);
+  char buf[64];
+
+  // List all the types so we get a compile warning if we've missed one,
+  // (using default: supresses the compiler warning.)
```

Comment:
I've listed all the types so I can ommit the default: statement. If a new type is added the compiler will warn it's not being handled rather than just let the default: swallow it.